File: static/CWE-502/csharp/patched.cs
// Remediation for CWE-502: Deserialization of Untrusted Data
// Fix: Deserialize with System.Text.Json into a known type instead of BinaryFormatter.

using System.Text.Json;

namespace GuineaPig.Sast
{
    class Patched_502
    {
        public class SafeDto
        {
            public string Name { get; set; }
            public int Value { get; set; }
        }

        public object LoadObject(byte[] userInput)
        {
            var options = new JsonSerializerOptions
            {
                PropertyNameCaseInsensitive = true
            };
            return JsonSerializer.Deserialize<SafeDto>(userInput, options);
        }
    }
}

File: static/CWE-502/java/patched.java
// Remediation for CWE-502: Deserialization of Untrusted Data
// Fix: Restrict ObjectInputStream to an explicit allowlist of expected classes.

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InvalidClassException;
import java.io.ObjectInputStream;
import java.io.ObjectStreamClass;
import java.util.Set;

class Patched_502 {
    private static final Set<String> ALLOWED = Set.of("java.lang.String", "java.lang.Number");

    static class SafeObjectInputStream extends ObjectInputStream {
        SafeObjectInputStream(ByteArrayInputStream in) throws IOException { super(in); }
        protected Class<?> resolveClass(ObjectStreamClass desc)
                throws IOException, ClassNotFoundException {
            if (!ALLOWED.contains(desc.getName())) {
                throw new InvalidClassException("Disallowed class", desc.getName());
            }
            return super.resolveClass(desc);
        }
    }

    public Object deserialize(byte[] userInput) throws Exception {
        try (ObjectInputStream ois =
                 new SafeObjectInputStream(new ByteArrayInputStream(userInput))) {
            return ois.readObject();
        }
    }
}
